crates/yttrium/src/sign/mod.rs

```diff
@@ -0,0 +1,647 @@
+// - handle connection error
+// - handle project ID/JWT error
+
+#[allow(unused)]
+impl Client {
+    pub fn new(
+        relay_url: String,
+        project_id: ProjectId,
+        client_id: ClientId,
+    ) -> Self {
+        Self { relay_url, project_id, client_id, websocket: None }
+    }
+
+    pub async fn _connect(&self) {
+        // TODO implement
+        // https://github.com/WalletConnect/walletconnect-monorepo/blob/5bef698dcf0ae910548481959a6a5d87eaf7aaa5/packages/sign-client/src/controllers/engine.ts#L220
+        unimplemented!()
+
+        // TODO call `wc_proposeSession`
+    }
+
+    pub async fn pair(&mut self, uri: &str) -> Result<Pairing, PairError> {
+        // TODO implement
+        // https://github.com/WalletConnect/walletconnect-monorepo/blob/5bef698dcf0ae910548481959a6a5d87eaf7aaa5/packages/sign-client/src/controllers/engine.ts#L330
+
+        // TODO parse URI
+        // https://github.com/WalletConnect/walletconnect-monorepo/blob/5bef698dcf0ae910548481959a6a5d87eaf7aaa5/packages/core/src/controllers/pairing.ts#L132
+        let pairing_uri = pairing_uri::parse(uri)
+            .map_err(|e| PairError::Internal(e.to_string()))?;
+
+        // TODO immediately throw if expired - maybe not necessary if FetchMessages returns empty array?
+        // note: no activatePairing
+        // TODO save symkey, if necessary
+
+        // TODO update relay method to not remove message
+
+        let response = self
+            .request::<FetchResponse>(relay_rpc::rpc::Params::FetchMessages(
+                FetchMessages { topic: pairing_uri.topic.clone() },
+            ))
+            .await
+            .map_err(|e| PairError::Internal(e.to_string()))?;
+
+        for message in response.messages {
+            if message.topic == pairing_uri.topic {
+                let decoded =
+                    BASE64.decode(message.message.as_bytes()).map_err(|e| {
+                        PairError::Internal(format!(
+                            "Failed to decode message: {e}"
+                        ))
+                    })?;
+                let envelope =
+                    envelope_type0::deserialize_envelope_type0(&decoded)
+                        .map_err(|e| PairError::Internal(e.to_string()))?;
+                let key = ChaCha20Poly1305::new(&pairing_uri.sym_key.into());
+                let decrypted = key
+                    .decrypt(&Nonce::from(envelope.iv), envelope.sb.as_slice())
+                    .map_err(|e| PairError::Internal(e.to_string()))?;
+                let request = serde_json::from_slice::<
+                    json_rpc::Request<serde_json::Value>,
+                >(&decrypted)
+                .map_err(|e| {
+                    PairError::Internal(format!(
+                        "Failed to parse decrypted message: {e}"
+                    ))
+                })?;
+                if request.meta.method != "wc_sessionPropose" {
```

Comment:
i would suggest to have a single place where we decrypt and then emit the request to designated function. Otherwise we will be decrypting a single request in multiple places just to check that method doesn't match. Example session request would be decrypted here for no reason.